libs/common/Networking/TcpNetworkHandlerBase.cs

```diff
@@ -144,22 +151,37 @@ void RecvEventArg_Completed(object sender, SocketAsyncEventArgs e)
                     {
                         // No more things to receive
                         Dispose(e);
-                        break;
+                        return;
                     }
-                    OnNetworkReceive(e.BytesTransferred);
-                    e.SetBuffer(networkReceiveBuffer, networkBytesRead, networkReceiveBuffer.Length - networkBytesRead);
-                } while (!e.AcceptSocket.ReceiveAsync(e));
+                    var receiveTask = OnNetworkReceiveAsync(e.BytesTransferred);
+                    if (!receiveTask.IsCompleted)
+                    {
+                        await receiveTask;
+                    }
+                    NetworkBufferProcessed(sender, e);
+                }
```

Comment:
nit: `while` should be on same line as `}` to maintain parity with style on `main`:

```
  do
  {
  } while (!e.AcceptSocket.ReceiveAsync(e));
```